session/session.go

```diff
@@ -1180,6 +1181,16 @@ func (s *session) ExecuteStmt(ctx context.Context, stmtNode ast.StmtNode) (sqlex
 
 // runStmt executes the sqlexec.Statement and commit or rollback the current transaction.
 func runStmt(ctx context.Context, se *session, s sqlexec.Statement) (rs sqlexec.RecordSet, err error) {
+	defer func() {
+		r := recover()
+		if r != nil {
+			if str, ok := r.(string); !ok || !strings.HasPrefix(str, memory.PanicMemoryExceed) {
+				panic(r)
+			}
+			err = errors.Errorf("%v", r)
```

Comment:
We should use 

```go
if !s.IsReadOnly(sessVars) && se.txn.Valid() {
	se.StmtRollback()
}
```
to keep transaction states clean.

We also need a `recover` in https://github.com/pingcap/tidb/blob/d941ff5cc8b4babf9dcfdd91b66a5c53b798c122/session/txn.go#L513-L523. The logical is.

```go
func (s *session) StmtCommit(memTracker *memory.Tracker) (err error) {
	defer func() {
		if r := recover(); r != nil {
			err = errors.Errorf("%v", r)
			st.doNotCommit = err
		}
		if err == nil && memTracker != nil {
			memTracker.Consume(int64(-s.txn.Size()))
		}
		s.txn.cleanup()
	}()
// Others...
}
```

Reply:
PTAL @tiancaiamao

Reply:
If we recover in `stmtCommit`, it seems that we do not need to recover in `runStmt`.